src/pdf_parser/pdf_utils/parsing_utils.py

```diff
@@ -575,7 +505,41 @@ def _calc_frac_overlap(block_1: lp.TextBlock, block_2: lp.TextBlock) -> float:
         intersection_width = block_1.intersect(block_2).width
         return intersection_width / union_width
 
-    def _infer_column_groups(self, threshold: float = 0.95):
+    @staticmethod
+    def _infer_missing_blocks_from_gaps(
+        column_blocks: lp.Layout, height_threshold: int = 50
+    ) -> List[lp.TextBlock]:
+        """Infer the missing blocks in columns by checking for sufficiently large gaps..
+
+        Args:
+            column_blocks: The text blocks to infer the missing blocks from.
+            height_threshold: The number of pixels for a missing blocks to be inferred (avoiding normal whitespace).
+
+        Returns:
+            The text blocks with the inferred missing blocks.
+        """
+        # Make sure the blocks are sorted by y_1.
+        column_blocks = column_blocks.sort(key=lambda b: b.coordinates[1])
+        x1 = min([b.coordinates[0] for b in column_blocks])
+        x2 = max([b.coordinates[2] for b in column_blocks])
+        # Iteratively fill in gaps between subsequent blocks.
+        new_blocks = []
+        for ix, block in enumerate(column_blocks):
+            for ix2, block2 in enumerate(column_blocks):
```

Comment:
it might be nicer to iterate through `idx in range(len(column_blocks)+1)` rather than use nested for loops here if you're only ever looking at neighbouring blocks (which i think you are)